### ISSUE TITLE
"Saving" current state?

### ISSUE THREAD

**Issue by niansa** [ggerganov/llama.cpp#202]
Hey!

Is it possible to add a way of dumping the current state into a file, so it can then be reloaded later? This would avoid the time needed to reload a long prompt over and over again.

Thanks
Niansa

**Comment by bitRAKE** [comment]
#174 also asked this, or do you have something else in mind?

**Comment by jart** [comment]
Thank you for using llama.cpp and thank you for sharing your feature request! You'll be excited to hear that what you're requesting is my top priority right now. I'm using #91 as the best place to discuss this, since the solution will entail using mmap(). Everyone is welcome to participate in helping us find the best solution. I believe mmap() will reduce startup latency to effectively zero, for everyone, and it'll work on nearly every platform on earth, including Windows, which has a nearly equivalent API.

**Comment by j-f1** [comment]
I think this is a different issue — that one is about changing how the model is loaded, this one is about reducing the time needed to reload a piece of text into the model.

**Comment by jart** [comment]
As you wish. Re-opening.

**Comment by niansa** [comment]
> #174 also asked this, or do you have something else in mind?

Basically yes, except that interactive user input and generated results should be saved too. So basically you can save and stop and just continue where the model/you've left off later or on another PC.

**Comment by bitRAKE** [comment]
I can't find it now, but @ggerganov said save/restore of the k&v tensors would preserve the state, iirc.
https://github.com/ggerganov/llama.cpp/blob/721311070e31464ac12bef9a4444093eb3eaebf7/main.cpp#L79-L82

**Comment by jarcen** [comment]
@bitRAKE Yes, those transformer's hidden state, preserving them is sufficient. Now, the question is how to edit them properly.
I'm also interested in removing n first elements to deal with context memory filling up.

**Comment by ggerganov** [comment]
Yes, it is the same
